File: ios/exelbid_plugin/Sources/exelbid_plugin/Mappers/AdOptionsMapper.swift
import CoreLocation
import ExelBidSDK

enum AdOptionsMapper {
    /// Decodes the EBAdOptions payload sent from Dart.
    /// Expected map shape:
    /// {
    ///   "keywords": {String: String},
    ///   "yearOfBirth": Int,        // 0 means unspecified
    ///   "gender": Int,             // 0 unspecified / 1 male / 2 female
    ///   "location": {"latitude": Double, "longitude": Double}?,
    ///   "coppa": Bool,
    ///   "testing": Bool,
    ///   "videoSkipMin": Int?,
    ///   "videoSkipAfter": Int?
    /// }
    static func decode(_ map: [String: Any]?) -> EBAdOptions {
        let options = EBAdOptions()
        guard let map = map else { return options }

        if let kw = map["keywords"] as? [String: String] {
            options.keywords = kw
        }
        if let year = map["yearOfBirth"] as? Int {
            options.yearOfBirth = year
        }
        if let genderRaw = map["gender"] as? Int,
           let gender = Gender(rawValue: genderRaw) {
            options.gender = gender
        }
        if let loc = map["location"] as? [String: Double],
           let lat = loc["latitude"], let lon = loc["longitude"] {
            options.location = CLLocation(latitude: lat, longitude: lon)
        }
        if let coppa = map["coppa"] as? Bool {
            options.coppa = coppa
        }
        if let testing = map["testing"] as? Bool {
            options.testing = testing
        }
        if let skipMin = map["videoSkipMin"] as? Int {
            options.videoSkipMin = skipMin
        }
        if let skipAfter = map["videoSkipAfter"] as? Int {
            options.videoSkipAfter = skipAfter
        }
        return options
    }
}

File: android/src/main/kotlin/com/exelbid/flutter/mappers/AdOptionsMapper.kt
package com.exelbid.flutter.mappers

import android.location.Location
import com.onnuridmc.exelbid.a

/**
 * Applies the `options` map sent from Dart (`AdOptions.toMap()`) to an ExelBid
 * in-house ad object. All ExelBid ad classes (`ExelBidAdView`,
 * `ExelBidInterstitial`, `ExelBidNative`) share the base class
 * `com.onnuridmc.exelbid.a`, which exposes the targeting setters.
 *
 * Map shape (mirrors iOS `AdOptionsMapper`):
 * ```
 * keywords: {String: String}, yearOfBirth: Int (0=unspecified),
 * gender: Int (0 unspecified / 1 male / 2 female),
 * location: {latitude: Double, longitude: Double}?, coppa: Bool, testing: Bool,
 * videoSkipMin: Int?, videoSkipAfter: Int?
 * ```
 *
 * Note: the Android SDK has no `videoSkipMin`/`videoSkipAfter` equivalent, so
 * those are ignored.
 */
object AdOptionsMapper {

    fun apply(target: a, options: Map<String, Any?>?) {
        options ?: return

        (options["testing"] as? Boolean)?.let { target.setTestMode(it) }
        (options["coppa"] as? Boolean)?.let { target.setCoppa(it) }

        (options["yearOfBirth"] as? Number)?.toInt()?.let {
            if (it > 0) target.setYob(it.toString())
        }

        // 0 unspecified → leave unset; 1 male → true; 2 female → false.
        when ((options["gender"] as? Number)?.toInt()) {
            1 -> target.setGender(true)
            2 -> target.setGender(false)
            else -> {}
        }

        (options["keywords"] as? Map<*, *>)?.forEach { (k, v) ->
            if (k is String && v is String) target.addKeyword(k, v)
        }

        (options["location"] as? Map<*, *>)?.let { loc ->
            val lat = (loc["latitude"] as? Number)?.toDouble()
            val lng = (loc["longitude"] as? Number)?.toDouble()
            if (lat != null && lng != null) {
                target.setLocation(
                    Location("flutter").apply {
                        latitude = lat
                        longitude = lng
                    },
                )
            }
        }
    }
}
